File: sdks/node/src/plugins/kiroide-installation.ts
import path from 'node:path';
import type { InstallConfig } from './base.js';
import { sameAgentId } from './common.js';
import {
  AGENT_KEY,
  AUDIT_SCRIPT,
  GUARD_SCRIPT,
  type KiroIdePaths,
  type RenderedKiroIdeDocument,
} from './kiroide-contract.js';
import {
  type KiroIdeSources,
  type LegacyKiroIdeDocument,
  requirePhysicalLegacyDirectory,
} from './kiroide-io.js';
import {
  commitManagedInstallation,
  preflightManagedInstallation,
  prepareManagedInstallation,
  type ManagedRuntimePaths,
  type PreparedManagedInstallation,
  type RenameFile,
} from './managed-installation.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
  type ManagedFileChange,
  type PreparedManagedTransaction,
} from './managed-transaction.js';

const DISPLAY_NAME = 'Kiro IDE';
const DIRECTORY_LABEL = 'Kiro IDE hooks directory';

function hookLocation(paths: KiroIdePaths) {
  return [{ directoryLabel: DIRECTORY_LABEL, filePath: paths.configPath }];
}

function removesLegacy(legacy: LegacyKiroIdeDocument, agentId?: string): boolean {
  return legacy.contract !== undefined
    && (agentId === undefined || sameAgentId(legacy.contract.agentId, agentId));
}

async function prepareLegacyRemoval(
  legacy: LegacyKiroIdeDocument,
  agentId?: string,
): Promise<ManagedFileChange | undefined> {
  if (!removesLegacy(legacy, agentId)) return undefined;
  if (legacy.raw === undefined) throw new Error(`Legacy Kiro IDE hook source is missing: ${legacy.filePath}`);
  await requirePhysicalLegacyDirectory(legacy);
  return prepareManagedFileChange({
    filePath: legacy.filePath,
    label: 'legacy Kiro IDE hook',
    next: undefined,
    mode: 0o600,
    expectedSource: legacy.raw,
    verifyExpectedSource: true,
  });
}

export async function preflightKiroIdeInstallation(
  config: InstallConfig,
  paths: KiroIdePaths,
): Promise<ManagedRuntimePaths> {
  return preflightManagedInstallation({
    agentKey: AGENT_KEY,
    hookLocations: hookLocation(paths),
    config,
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
}

export async function prepareKiroIdeInstallation(
  config: InstallConfig,
  sources: KiroIdeSources,
  rendered: RenderedKiroIdeDocument,
): Promise<PreparedManagedInstallation> {
  if (rendered.next === undefined) throw new Error('Kiro IDE installation requires a hook document');
  const prepared = await prepareManagedInstallation({
    agentKey: AGENT_KEY,
    displayName: DISPLAY_NAME,
    hookSources: [{
      directoryLabel: DIRECTORY_LABEL,
      label: 'Kiro IDE hooks',
      filePath: sources.paths.configPath,
      expectedSource: sources.document.raw,
      source: rendered.next,
    }],
    config,
    auditOptions: { nativePayload: true },
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
  const legacyChange = await prepareLegacyRemoval(sources.legacy, config.agentId);
  if (!legacyChange) return prepared;
  return {
    ...prepared,
    transaction: {
      ...prepared.transaction,
      directories: [
        ...prepared.transaction.directories,
        { path: path.dirname(sources.legacy.filePath), label: 'legacy Kiro IDE hooks directory' },
      ],
      changes: [...prepared.transaction.changes, legacyChange],
    },
  };
}

export async function commitKiroIdeInstallation(
  prepared: PreparedManagedInstallation,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedInstallation(prepared, renameFile);
}

export async function prepareKiroIdeUninstall(
  sources: KiroIdeSources,
  rendered: RenderedKiroIdeDocument,
  agentId?: string,
): Promise<PreparedManagedTransaction> {
  const changes: ManagedFileChange[] = [];
  const directories = [];
  if (rendered.changed) {
    const configChange = await prepareManagedFileChange({
      filePath: sources.document.filePath,
      label: 'Kiro IDE hooks',
      next: rendered.next,
      mode: 0o600,
      expectedSource: sources.document.raw,
      verifyExpectedSource: true,
    });
    if (configChange) {
      changes.push(configChange);
      directories.push({ path: sources.paths.hooksDirectory, label: DIRECTORY_LABEL });
    }
  }
  const legacyChange = await prepareLegacyRemoval(sources.legacy, agentId);
  if (legacyChange) {
    changes.push(legacyChange);
    directories.push({
      path: path.dirname(sources.legacy.filePath),
      label: 'legacy Kiro IDE hooks directory',
    });
  }
  return {
    displayName: DISPLAY_NAME,
    operation: 'uninstall',
    directories,
    changes,
  };
}

export async function commitKiroIdeUninstall(
  prepared: PreparedManagedTransaction,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedTransaction(prepared, renameFile);
}

File: sdks/go/cmd/elydora/plugins/kiroide_installation.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
)

type kiroIdeRuntimePaths struct {
	managedRuntimePaths
	initialDirectories []kiroIdeDirectoryIdentity
}

type kiroIdePreparedTransaction struct {
	operation          string
	paths              *kiroIdePaths
	runtimePaths       *kiroIdeRuntimePaths
	changes            []*fileChange
	preconditions      []filePrecondition
	removeLegacy       bool
	initialDirectories []kiroIdeDirectoryIdentity
}

func preflightKiroIdeInstallation(
	config InstallConfig,
) (*kiroIdeRuntimePaths, string, error) {
	if err := validateManagedInstallConfig(config, kiroIdeAgentKey, "Kiro IDE"); err != nil {
		return nil, "", err
	}
	managed, err := resolveManagedRuntimePaths(config, kiroIdeGuardScript, kiroIdeAuditScript)
	if err != nil {
		return nil, "", err
	}
	paths := &kiroIdeRuntimePaths{managedRuntimePaths: *managed}
	initialDirectories := make([]kiroIdeDirectoryIdentity, 0, 3)
	for _, item := range []struct{ path, label string }{
		{filepath.Dir(paths.runtimeRoot), "home directory"},
		{paths.runtimeRoot, "Elydora runtime directory"},
		{paths.agentDirectory, "Elydora agent runtime directory"},
	} {
		identity, err := snapshotKiroIdeDirectory(item.path, item.label)
		if err != nil {
			return nil, "", err
		}
		initialDirectories = append(initialDirectories, identity)
	}
	if initialDirectories[0].info == nil {
		return nil, "", fmt.Errorf(
			"home directory is missing: %s",
			initialDirectories[0].path,
		)
	}
	if err := validateManagedRuntimeIdentity(
		paths.agentDirectory, config.AgentID, kiroIdeAgentKey, "Kiro IDE",
	); err != nil {
		return nil, "", err
	}
	if err := assertKiroIdeDirectoryStates(
		initialDirectories,
		"Kiro IDE runtime preflight",
	); err != nil {
		return nil, "", err
	}
	paths.initialDirectories = initialDirectories
	nodePath, err := resolveNodeRuntime()
	if err != nil {
		return nil, "", err
	}
	return paths, nodePath, nil
}

func prepareKiroIdeFile(
	filePath string,
	label string,
	contents []byte,
	mode os.FileMode,
	maximumBytes int64,
) (*fileChange, *filePrecondition, error) {
	snapshot, err := readManagedFile(filePath, label, maximumBytes)
	if err != nil {
		return nil, nil, err
	}
	change, err := prepareSnapshotSourceChange(
		filePath,
		label,
		snapshot,
		contents,
		mode,
		false,
	)
	if err != nil {
		return nil, nil, err
	}
	if change != nil {
		return change, nil, nil
	}
	condition := &filePrecondition{
		filePath: filePath, label: label, snapshot: snapshot,
		maximumSize: maximumBytes,
	}
	return nil, condition, nil
}

func appendKiroIdeChange(
	prepared *kiroIdePreparedTransaction,
	change *fileChange,
	condition *filePrecondition,
) {
	if change != nil {
		prepared.changes = append(prepared.changes, change)
	}
	if condition != nil {
		prepared.preconditions = append(prepared.preconditions, *condition)
	}
}

func prepareKiroIdeInstallation(
	config InstallConfig,
	sources *kiroIdeSources,
	paths *kiroIdeRuntimePaths,
	rendered *kiroIdeRenderedDocument,
) (*kiroIdePreparedTransaction, error) {
	if sources == nil || paths == nil || rendered == nil {
		return nil, fmt.Errorf("Kiro IDE installation requires prepared sources")
	}
	runtimeConfig, err := buildManagedRuntimeConfig(config, kiroIdeAgentKey)
	if err != nil {
		return nil, err
	}
	guardScript, auditScript := kiroIdeExpectedScripts(config.AgentID, nil)
	prepared := &kiroIdePreparedTransaction{
		operation: "installation", paths: sources.paths, runtimePaths: paths,
		initialDirectories: append(
			append([]kiroIdeDirectoryIdentity(nil), sources.initialDirectories...),
			paths.initialDirectories...,
		),
	}
	for _, item := range []struct {
		path, label string
		contents    []byte
		mode        os.FileMode
		maximum     int64
	}{
		{paths.guardPath, "Elydora guard runtime", guardScript, 0700, maxManagedSourceBytes},
		{paths.configPath, "Elydora runtime config", runtimeConfig, 0600, maxRuntimeConfigBytes},
		{paths.keyPath, "Elydora private key", []byte(config.PrivateKey), 0600, maxProtectedSecretBytes},
		{paths.auditPath, "Elydora audit runtime", auditScript, 0700, maxManagedSourceBytes},
	} {
		change, condition, err := prepareKiroIdeFile(
			item.path, item.label, item.contents, item.mode, item.maximum,
		)
		if err != nil {
			return nil, err
		}
		appendKiroIdeChange(prepared, change, condition)
	}
	documentChange, err := prepareRenderedKiroIdeChange(rendered)
	if err != nil {
		return nil, err
	}
	if documentChange == nil {
		prepared.preconditions = append(prepared.preconditions, filePrecondition{
			filePath: sources.document.filePath, label: "Kiro IDE hooks",
			snapshot: sources.document.snapshot,
		})
	} else {
		prepared.changes = append(prepared.changes, documentChange)
	}
	removeLegacy := sources.legacy.contract != nil &&
		sameManagedAgentID(sources.legacy.contract.agentID, config.AgentID)
	if removeLegacy {
		legacyChange, err := prepareSnapshotSourceChange(
			sources.legacy.filePath,
			"legacy Kiro IDE hook",
			sources.legacy.snapshot,
			nil,
			0600,
			true,
		)
		if err != nil {
			return nil, err
		}
		if legacyChange == nil {
			return nil, fmt.Errorf(
				"legacy Kiro IDE hook disappeared before migration: %s",
				sources.legacy.filePath,
			)
		}
		prepared.changes = append(prepared.changes, legacyChange)
		prepared.removeLegacy = true
	} else {
		prepared.preconditions = append(prepared.preconditions, filePrecondition{
			filePath: sources.legacy.filePath, label: "legacy Kiro IDE hook",
			snapshot: sources.legacy.snapshot,
		})
	}
	if err := assertKiroIdeDirectoryStates(
		prepared.initialDirectories,
		"Kiro IDE installation preparation",
	); err != nil {
		return nil, err
	}
	return prepared, nil
}

func prepareKiroIdeUninstall(
	sources *kiroIdeSources,
	rendered *kiroIdeRenderedDocument,
	agentID string,
) (*kiroIdePreparedTransaction, error) {
	prepared := &kiroIdePreparedTransaction{
		operation: "uninstall", paths: sources.paths,
		initialDirectories: append(
			[]kiroIdeDirectoryIdentity(nil),
			sources.initialDirectories...,
		),
	}
	documentChange, err := prepareRenderedKiroIdeChange(rendered)
	if err != nil {
		return nil, err
	}
	if documentChange == nil {
		prepared.preconditions = append(prepared.preconditions, filePrecondition{
			filePath: sources.document.filePath, label: "Kiro IDE hooks",
			snapshot: sources.document.snapshot,
		})
	} else {
		prepared.changes = append(prepared.changes, documentChange)
	}
	removeLegacy := sources.legacy.contract != nil &&
		(agentID == "" || sameManagedAgentID(sources.legacy.contract.agentID, agentID))
	if removeLegacy {
		change, err := prepareSnapshotSourceChange(
			sources.legacy.filePath,
			"legacy Kiro IDE hook",
			sources.legacy.snapshot,
			nil,
			0600,
			true,
		)
		if err != nil {
			return nil, err
		}
		if change != nil {
			prepared.changes = append(prepared.changes, change)
			prepared.removeLegacy = true
		}
	} else {
		prepared.preconditions = append(prepared.preconditions, filePrecondition{
			filePath: sources.legacy.filePath, label: "legacy Kiro IDE hook",
			snapshot: sources.legacy.snapshot,
		})
	}
	if err := assertKiroIdeDirectoryStates(
		prepared.initialDirectories,
		"Kiro IDE uninstall preparation",
	); err != nil {
		return nil, err
	}
	return prepared, nil
}

func commitKiroIdeTransaction(
	prepared *kiroIdePreparedTransaction,
	rename renameFunc,
) error {
	if prepared == nil {
		return fmt.Errorf("Kiro IDE transaction is missing")
	}
	operation := "Kiro IDE " + prepared.operation
	if err := assertKiroIdeDirectoryStates(
		prepared.initialDirectories,
		operation,
	); err != nil {
		return err
	}
	workspaceChange := false
	for _, change := range prepared.changes {
		if change != nil && sameManagedPath(change.filePath, prepared.paths.configPath) {
			workspaceChange = true
			break
		}
	}
	directories := make([]kiroIdeDirectoryIdentity, 0, len(prepared.initialDirectories))
	for _, directory := range prepared.initialDirectories {
		if directory.info != nil {
			directories = appendKiroIdeDirectory(directories, directory)
		}
	}
	establish := func(path string, private bool) error {
		if err := assertKiroIdeDirectoryStates(directories, operation); err != nil {
			return err
		}
		initial, err := kiroIdeInitialDirectory(prepared, path)
		if err != nil {
			return err
		}
		current, err := establishKiroIdeDirectory(initial, operation, private)
		if err != nil {
			return err
		}
		directories = appendKiroIdeDirectory(directories, current)
		return assertKiroIdeDirectoryStates(directories, operation)
	}
	if workspaceChange {
		if err := establish(prepared.paths.kiroDirectory, false); err != nil {
			return err
		}
		if err := establish(prepared.paths.hooksDirectory, false); err != nil {
			return err
		}
	}
	if prepared.runtimePaths != nil {
		if err := establish(prepared.runtimePaths.runtimeRoot, true); err != nil {
			return err
		}
		if err := establish(prepared.runtimePaths.agentDirectory, true); err != nil {
			return err
		}
	}
	preconditions := append([]filePrecondition(nil), prepared.preconditions...)
	preconditions = append(preconditions, kiroIdeDirectoryPreconditions(directories)...)
	return writeChangesWithFileOps(
		prepared.changes,
		operation,
		guardedKiroIdeTransactionOps(rename, directories),
		preconditions...,
	)
}

func commitKiroIdeInstallation(
	prepared *kiroIdePreparedTransaction,
	rename renameFunc,
) error {
	return commitKiroIdeTransaction(prepared, rename)
}

func commitKiroIdeUninstall(
	prepared *kiroIdePreparedTransaction,
	rename renameFunc,
) error {
	return commitKiroIdeTransaction(prepared, rename)
}
